cql3/restrictions/statement_restrictions.cc

```diff
@@ -1057,6 +1061,359 @@ ::shared_ptr<expression> make_conjunction(::shared_ptr<expression> a, ::shared_p
+                } else if (opr.op == operator_type::GT || opr.op == operator_type::GTE) {
+                    return true;
+                } else {
+                    throw std::logic_error("Unknown slice operator");
+                }
+            }
+            // Otherwise, we don't know the LHS/RHS order, so check the next component.
+        }
+        // Getting here means LHS == RHS.
+        return opr.op == operator_type::LTE || opr.op == operator_type::GTE;
+    } else if (columns.size() == 1) {
+        auto lhs = get_value(columns[0], selection, cells, options);
+        if (!lhs) {
+            lhs = bytes(); // Compatible with old code, which feeds null to type comparators.
+        }
+        auto rhs = to_bytes_opt(opr.rhs->bind_and_get(options));
+        if (!rhs) {
+            return false;
+        }
+        return limits(*lhs, opr.op, *rhs, *columns[0].col->type);
+    } else {
+        throw std::logic_error("empty tuple on LHS of an inequality");
+    }
+}
+
+/// True iff collection (list, set, or map) contains value.
+bool contains(const data_value& collection, const raw_value_view& value) {
+    if (!value) {
+        return true;
+    }
+    auto col_type = static_pointer_cast<const collection_type_impl>(collection.type());
+    auto&& element_type = col_type->is_set() ? col_type->name_comparator() : col_type->value_comparator();
+    return with_linearized(*value, [&] (bytes_view val) {
+        auto exists_in = [&](auto&& range) {
+            auto found = std::find_if(range.begin(), range.end(), [&] (auto&& element) {
+                return element_type->compare(element.serialize_nonnull(), val) == 0;
+            });
+            return found != range.end();
+        };
+        if (col_type->is_list()) {
+            if (!exists_in(value_cast<list_type_impl::native_type>(collection))) {
+                return false;
+            }
+        } else if (col_type->is_set()) {
+            if (!exists_in(value_cast<set_type_impl::native_type>(collection))) {
+                return false;
+            }
+        } else {
+            auto data_map = value_cast<map_type_impl::native_type>(collection);
+            if (!exists_in(data_map | boost::adaptors::transformed([] (auto&& p) { return p.second; }))) {
+                return false;
+            }
+        }
+        return true;
+    });
+}
+
+/// True iff columns is a single collection containing value.
+bool contains(const raw_value_view& value, const std::vector<column_value>& columns, const selection& selection,
+              row_data cells) {
+    if (columns.size() != 1) {
+        throw exceptions::unsupported_operation_exception("tuple CONTAINS not allowed");
+    }
+    if (columns[0].sub) {
+        throw exceptions::unsupported_operation_exception("CONTAINS lhs is subscripted");
+    }
+    const auto collection = get_value(columns[0], selection, cells,
+                                      query_options::DEFAULT /*unused when .sub is null*/);
```

Comment:
It's not a good idea to pass query_options::DEFAULT, even if now it happens to be unused. It's asking for a bug to creep in later.

Reply:
Replaced with real options.